conmon-rs/server/src/init.rs

```diff
@@ -1,24 +1,172 @@
 use log::info;
-use std::fs::File;
-use std::io::{ErrorKind, Write};
+use std::{
+    fs::File,
+    io::{self, ErrorKind, Write},
+    path::Path,
+};
 
-/// Unset the locale for the current process.
-pub fn unset_locale() {
-    setlocale(LocaleCategory::LcAll, "");
+#[derive(Debug, Default)]
+pub struct Init<T> {
+    imp: T,
 }
 
-/// Helper to adjust the OOM score of the currently running process.
-pub fn set_oom(score: &str) -> Result<()> {
-    // Attempt adjustment with best-effort.
-    if let Err(err) = File::create("/proc/self/oom_score_adj")?.write_all(score.as_bytes()) {
-        match err.kind() {
-            ErrorKind::PermissionDenied => {
-                info!("Missing sufficient privileges to adjust OOM score")
+impl<T> Init<T>
+where
+    T: InitImpl,
+{
+    /// Unset the locale for the current process.
+    pub fn unset_locale(&mut self) {
+        self.imp.setlocale(LocaleCategory::LcAll, "");
+    }
+
+    /// Helper to adjust the OOM score of the currently running process.
+    pub fn set_oom<S: AsRef<str>>(&mut self, score: S) -> Result<()> {
+        // Attempt adjustment with best-effort.
+        let mut file = self.imp.create_file("/proc/self/oom_score_adj")?;
+        if let Err(err) = self
+            .imp
+            .write_all_file(&mut file, score.as_ref().as_bytes())
+        {
+            match err.kind() {
+                ErrorKind::PermissionDenied => {
+                    info!("Missing sufficient privileges to adjust OOM score")
+                }
+                _ => bail!("adjusting OOM score {}", err),
             }
-            _ => bail!("adjusting OOM score {}", err),
         }
+        Ok(())
+    }
+}
+
+pub trait InitImpl {
+    fn setlocale<T: Into<Vec<u8>>>(
+        &mut self,
+        category: LocaleCategory,
+        locale: T,
+    ) -> Option<Vec<u8>>;
+    fn create_file<P: AsRef<Path>>(&mut self, path: P) -> io::Result<File>;
+    fn write_all_file(&mut self, file: &mut File, buf: &[u8]) -> io::Result<()>;
+}
+
+#[derive(Debug, Default)]
+pub struct DefaultInit;
+
+impl InitImpl for DefaultInit {
+    fn setlocale<T: Into<Vec<u8>>>(
+        &mut self,
+        category: LocaleCategory,
+        locale: T,
+    ) -> Option<Vec<u8>> {
+        setlocale(category, locale)
+    }
+
+    fn create_file<P: AsRef<Path>>(&mut self, path: P) -> io::Result<File> {
+        File::create(path)
+    }
+
+    fn write_all_file(&mut self, file: &mut File, buf: &[u8]) -> io::Result<()> {
+        file.write_all(buf)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use anyhow::Context;
+    use std::{path::PathBuf, str};
+    use tempfile::tempfile;
+
+    #[derive(Default)]
+    struct TestInit {
+        set_locale_called: Option<LocaleCategory>,
+        create_file_called: Option<PathBuf>,
+        create_file_fails: bool,
+        write_all_called: Option<Vec<u8>>,
+        write_all_fails: Option<ErrorKind>,
+    }
+
+    impl InitImpl for TestInit {
```

Comment:
mockall crate seems like a good mock library... it would remove the boilerplate here.

Reply:
Thanks, changed as suggested. PTAL again

Reply:
wow... wasn't expecting it to be this clean, but it looks great.